data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/NewlineDelimitedCodec.java

```diff
@@ -42,18 +47,34 @@ public void parse(final InputStream inputStream, final Consumer<Record<Event>> e
     }
 
     private void parseBufferedReader(final BufferedReader reader, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        final boolean doAddHeaderToOutgoingEvents = Objects.nonNull(this.headerDestination);
+        boolean hasReadHeader = false;
+        String header = "";
+
         int linesToSkip = skipLines;
         String line;
         while ((line = reader.readLine()) != null) {
 
-            if (linesToSkip > 0) {
-                linesToSkip--;
+            if (linesToSkip > 0 || (doAddHeaderToOutgoingEvents && !hasReadHeader)) {
```

Comment:
I think if you split this clause into two separate `if` statements you could simplify this and remove some branching. As it is now it seems like there would be a bug if you wanted to skip multiple lines since `hasReadHeader` would be set to `true` on the first iteration of the loop

I'm thinking something like:
```
if (linesToSkip > 0) {
    linesToSkip--;
    continue;
}

final Map<String, String> eventData = new HashMap<>();

if (doAddHeaderToOutgoingEvents && !hasReadHeader) {
    header = line;
    hasReadHeader = true;
    eventData.put(this.headerDestination, header);
}

eventData.put(MESSAGE_FIELD_NAME, line);
final Event event = JacksonLog.builder().withData(eventData).build();

Reply:
Ignore the above suggestion I wasn't properly accounting for the `continue`. I think what's here is fine.

You might consider using some variables to break down the boolean logic and make things more readable. Something like:
```
final boolean stillNeedToReadHeader = doAddHeaderToOutgoingEvents && !hasReadHeader;
final boolean shouldSkipLine = linesToSkip > 0 || stillNeedToReadHeader;
if (shouldSkipLine) {
    ...
}
```


I think the end of the method could also be rewritten to make the code shorter and more declarative by getting rid of the `else` but that's just a personal preference. 
```
final Map<String, String> eventData = new HashMap<>(); 
if (doAddHeaderToOutgoingEvents) { 
    eventData.put(this.headerDestination, header); 
} 

eventData.put(MESSAGE_FIELD_NAME, line); 
final Event event = JacksonLog.builder().withData(eventData).build();
```

Reply:
Thanks for your feedback! I agree that the boolean logic needs some work. I took your suggestion and added some variables to make the first statement more readable. I also agree that getting rid of the `else` is more declarative and I've made that change. I'll push these changes to this branch along with the next review to avoid triggering a re-review.